tests/unit/s2n_client_hello_retry_test.c

```diff
@@ -692,6 +693,331 @@ int main(int argc, char **argv)
+            EXPECT_SUCCESS(s2n_connections_set_io_pair(client_conn, server_conn, &io_pair));
+
+            /* Force the HRR path */
+            client_conn->security_policy_override = &security_policy_test_tls13_retry;
+
+            /* Send ClientHello */
+            s2n_blocked_status blocked = 0;
+            EXPECT_OK(s2n_negotiate_until_message(client_conn, &blocked, SERVER_HELLO));
+
+            /* Clear server name.
+             * Without a server name, we don't send the server name extension. */
+            client_conn->server_name[0] = '\0';
+
+            /* Expect failure because second client hello doesn't match */
+            EXPECT_FAILURE_WITH_ERRNO(s2n_negotiate_test_server_and_client(server_conn, client_conn),
+                    S2N_ERR_BAD_MESSAGE);
+        }
+
+        /* Test: The server rejects a second ClientHello with an added extension */
+        {
+            DEFER_CLEANUP(struct s2n_connection *server_conn = s2n_connection_new(S2N_SERVER),
+                    s2n_connection_ptr_free);
+            EXPECT_NOT_NULL(server_conn);
+            EXPECT_SUCCESS(s2n_connection_set_blinding(server_conn, S2N_SELF_SERVICE_BLINDING));
+            EXPECT_SUCCESS(s2n_connection_set_config(server_conn, config));
+
+            DEFER_CLEANUP(struct s2n_connection *client_conn = s2n_connection_new(S2N_CLIENT),
+                    s2n_connection_ptr_free);
+            EXPECT_NOT_NULL(server_conn);
+            EXPECT_SUCCESS(s2n_connection_set_blinding(client_conn, S2N_SELF_SERVICE_BLINDING));
+            EXPECT_SUCCESS(s2n_connection_set_config(client_conn, config));
+
+            struct s2n_test_io_pair io_pair = { 0 };
+            EXPECT_SUCCESS(s2n_io_pair_init_non_blocking(&io_pair));
+            EXPECT_SUCCESS(s2n_connections_set_io_pair(client_conn, server_conn, &io_pair));
+
+            /* Force the HRR path */
+            client_conn->security_policy_override = &security_policy_test_tls13_retry;
+
+            /* Send ClientHello */
+            s2n_blocked_status blocked = 0;
+            EXPECT_OK(s2n_negotiate_until_message(client_conn, &blocked, SERVER_HELLO));
+
+            /* Add a server name.
+             * Without a server name, we don't send the server name extension. */
+            EXPECT_SUCCESS(s2n_set_server_name(client_conn, "localhost"));
+
+            /* Expect failure because second client hello doesn't match */
+            EXPECT_FAILURE_WITH_ERRNO(s2n_negotiate_test_server_and_client(server_conn, client_conn),
+                    S2N_ERR_BAD_MESSAGE);
+        }
+
+        /* Test: If the initial ClientHello includes all extensions, so does the second ClientHello.
+         *
+         * This includes TLS1.2 extensions, since the ClientHello is sent before
+         * the client knows what version the server will negotiate.
+         *
+         * We have to test with all extensions to ensure that an s2n server will never reject
```

Comment:
I'm annoying :)
```suggestion
         * We have to test with all extensions to ensure that a s2n server will never reject
```

Reply:
Completely unrelated to TLS or code, but this is actually an interesting quirk of English! "an" is correct here, because "s2n" is pronounced with a vowel sound at the beginning, because we pronounce the letter "s" as kind of "es". And the "a" vs "an" choice is based on sound, not letter. https://ell.stackexchange.com/questions/6760/should-individual-letters-be-preceded-with-an